### PR TITLE
[SYCLomatic] Provide option query-api-mapping.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

**Review by yihwang-nv — APPROVED** [review]
LGTM

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                                     const std::string &Description) {
+  EntryMap[Name] = Description;
+}
+
+void APIMappingEntrys::initEntryMap() {
+  static const std::string SYCL_API = "The API is migrated to SYCL API:\n  ";
+  static const std::string DPCT_API =
+      "The API is migrated to helper function:\n  ";
+
+#define REGISTER_ENTRY(NAME, DESC) registerEntry(NAME, DESC);
```
Comment (tangjj11) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: Need add a example of this macro. From @ziranzha.

File: clang/lib/DPCT/APINames.inc [hunk 2]
```diff
@@ -34,1905 +34,1905 @@
...
-ENTRY(cudaIpcCloseMemHandle, cudaIpcCloseMemHandle, true, NO_FLAG, P0, "DPCT1030")
-ENTRY(cudaIpcGetEventHandle, cudaIpcGetEventHandle, true, NO_FLAG, P0, "DPCT1030")
-ENTRY(cudaIpcGetMemHandle, cudaIpcGetMemHandle, true, NO_FLAG, P0, "DPCT1030")
-ENTRY(cudaIpcOpenEventHandle, cudaIpcOpenEventHandle, true, NO_FLAG, P0, "DPCT1030")
-ENTRY(cudaIpcOpenMemHandle, cudaIpcOpenMemHandle, true, NO_FLAG, P0, "DPCT1030")
-ENTRY(cudaSetDevice, cudaSetDevice, true, NO_FLAG, P0, "DPCT1093")
-ENTRY(cudaSetDeviceFlags, cudaSetDeviceFlags, true, NO_FLAG, P0, "DPCT1026/DPCT1027")
-ENTRY(cudaSetValidDevices, cudaSetValidDevices, false, NO_FLAG, P4, "comment")
-ENTRY(cudaInitDevice, cudaInitDevice, false, NO_FLAG, P4, "comment")
+ENTRY(cudaChooseDevice, cudaChooseDevice, false, NO_FLAG, P0, "comment", "NA")
```
Comment (wenhuiNi) [inline clang/lib/DPCT/APINames.inc:37]: please add a first true APIMapping(Not "NA") as a example

**Review by wenhuiNi — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/APINames.inc (hunk 2 above)
Comment (tangjj11) [reply clang/lib/DPCT/APINames.inc:37]: The last test case contain all the not "NA" API mapping.

**Review by wenhuiNi — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 3]
```diff
@@ -416,4 +416,11 @@ DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool>), GenHelperFunctio
         llvm::cl::desc("Generate helper function files in the --out-root directory. Default: off."),
         llvm::cl::cat(DPCTCat), llvm::cl::init(false))
 
+DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<std::string>),
+                     QueryAPIMapping, "query-api-mapping",
+                     llvm::cl::desc("Outputs to stdout functionally compatible "
+                                    "SYCL API mapping for CUDA API."),
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: Query functionally compatible SYCL API to migrate CUDA API.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+#define DPCT_QUERY_API_MAPPING_H
+
+#include <string>
+#include <unordered_map>
+
+namespace clang {
+namespace dpct {
+
+class APIMappingEntrys {
+  std::string Name;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: L19, and L20  not used?

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+#ifndef DPCT_QUERY_API_MAPPING_H
+#define DPCT_QUERY_API_MAPPING_H
+
+#include <string>
+#include <unordered_map>
+
+namespace clang {
+namespace dpct {
+
+class APIMappingEntrys {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.h]:
```suggestion
class APIMapping{
```

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  static std::unordered_map<std::string, std::string> EntryMap;
+
+  static void registerEntry(const std::string &Name,
+                            const std::string &Description);
+
+public:
+  static void initEntryMap();
+
+  template <class StreamTy>
+  static void printMappingDesc(StreamTy &Stream, const std::string &Key) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.h]:
```suggestion
  static void queryAPIMapping(StreamTy &Stream, const std::string &Key) {
```

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+void APIMappingEntrys::registerEntry(const std::string &Name,
+                                     const std::string &Description) {
+  EntryMap[Name] = Description;
+}
+
+void APIMappingEntrys::initEntryMap() {
+  static const std::string MultiStr =
+      "\nThere are multi migration solutions for this API with different "
+      "migration options,\nsuggest to use the tool to migrate a complete test "
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: a complete test case  ==> a API usage code

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+std::unordered_map<std::string, std::string> APIMappingEntrys::EntryMap;
+
+void APIMappingEntrys::registerEntry(const std::string &Name,
+                                     const std::string &Description) {
+  EntryMap[Name] = Description;
+}
+
+void APIMappingEntrys::initEntryMap() {
+  static const std::string MultiStr =
+      "\nThere are multi migration solutions for this API with different "
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]:
```suggestion
      "\nThere are multi kinds of migrations for this API with different "
```

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      "migration options,\nsuggest to use the tool to migrate a complete test "
+      "case to see more detail of the migration.";
+#define ENTRY(INTERFACENAME, APINAME, VALUE, FLAG, TARGET, COMMENT, MAPPING)   \
+  registerEntry(#APINAME, MAPPING);
+#define ENTRY_MEMBER_FUNCTION(OBJNAME, INTERFACENAME, APINAME, VALUE, FLAG,    \
+                              TARGET, COMMENT, MAPPING)                        \
+  registerEntry(#OBJNAME "." #APINAME, MAPPING);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp:32]: so it's runtime to build the map,  can we make the map build time available.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+void APIMappingEntrys::initEntryMap() {
+  static const std::string MultiStr =
+      "\nThere are multi migration solutions for this API with different "
+      "migration options,\nsuggest to use the tool to migrate a complete test "
+      "case to see more detail of the migration.";
+#define ENTRY(INTERFACENAME, APINAME, VALUE, FLAG, TARGET, COMMENT, MAPPING)   \
+  registerEntry(#APINAME, MAPPING);
+#define ENTRY_MEMBER_FUNCTION(OBJNAME, INTERFACENAME, APINAME, VALUE, FLAG,    \
+                              TARGET, COMMENT, MAPPING)                        \
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp:31]: and if mapping if "NA", no need to build into the map.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 9 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp:32]: The initial function only be called when user use --query-api-mapping option. So in most user scenarios, this map is empty and will save more memory when use dpct to migrate code.
When user use --query-api-mapping option, dpct will build the map and returns (without migrate code), I think it is fast enough.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 10 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp:31]: fixed

File: clang/include/clang/DPCT/DPCTOptions.inc (hunk 3 above)
Comment (tangjj11) [reply clang/include/clang/DPCT/DPCTOptions.inc]: fixed

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h (hunk 4 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: deleted

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h (hunk 5 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: fixed

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h (hunk 6 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: fixed

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 7 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: fixed

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 8 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**